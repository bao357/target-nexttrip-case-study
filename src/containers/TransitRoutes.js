import React from 'react'
import { connect } from 'react-redux'
import { setTransitRoute, getTransitRoutes } from '../actions'
import Error from './Error'
import Loading from './Loading'
import WizardNavigation from './WizardNavigation'

class TransitRoutes extends React.Component {

    componentDidMount() {
        this.props.getTransitRoutes()
    }

    componentDidUpdate() {

        //set state if they dont update dropdown
        //TODO: move to saga to set state intially
        if (!this.props.direction) {
            this.props.setTransitRoute(this.getCurrentSelectedValue())
        }
    }

    getCurrentSelectedValue = () => {
        return document.getElementById("transit-route-dropdown").value
    }

    handleOnChange = (event) => {
        this.props.setTransitRoute(this.getCurrentSelectedValue())
    }

    render() {

        let transitRouteList = []

        if (this.props && this.props.transitRoutes) {

            transitRouteList = this.props.transitRoutes.map((item) => {
                return (<option key={item.Route} value={item.Route}>{item.Description}</option>)
            })
        }

        return (
            <div className="transit-routes-route-page">
                <Error />
                <Loading />
                <div className="field">
                    <label className="label">Select your transit route:</label>
                    <div className="control">
                        <div className="select">
                            <select id="transit-route-dropdown"
                                defaultValue={this.props.transitRoute}
                                onChange={this.handleOnChange}
                            >
                                {transitRouteList}
                            </select>
                        </div>
                    </div>
                </div>
                <WizardNavigation />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    transitRoutes: state.transitRoutes,
    transitRoute: state.transitRoute
})

const mapDispatchToProps = (dispatch) => ({
    getTransitRoutes: () => { dispatch(getTransitRoutes()) },
    setTransitRoute: (transitRoute) => { dispatch(setTransitRoute(transitRoute)) }
})

TransitRoutes = connect(mapStateToProps, mapDispatchToProps)(TransitRoutes)
export default TransitRoutes