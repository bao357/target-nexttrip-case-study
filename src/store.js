import createSagaMiddleware from 'redux-saga'
import { createStore, applyMiddleware } from 'redux'
import rootSaga from './sagas'
import reducer from './reducers'

const sagaMiddleware = createSagaMiddleware()
const store = createStore(
    reducer,
    applyMiddleware(sagaMiddleware),
)

sagaMiddleware.run(rootSaga)

export default store