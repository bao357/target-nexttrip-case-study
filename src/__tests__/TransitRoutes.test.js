import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'
import { Provider } from 'react-redux'
import { BrowserRouter as Router } from 'react-router-dom'

import store from '../store'
import TransitRoutes from '../containers/TransitRoutes'

describe('Transit routes', () => {
    let container
    beforeEach(() => {
        container = document.createElement('div')
        document.body.appendChild(container)
    })

    afterEach(() => {
        document.body.removeChild(container)
        container = null
    })

    function render() {
        act(() => {
            ReactDOM.render(
                <Provider store={store}>
                    <Router>
                        <TransitRoutes />
                    </Router>
                </Provider>,
                container
            )
        })
    }

    it('should render the Transit Route page with loading initially', () => {
        render()
        expect(document.body).toMatchSnapshot()
        expect(document.getElementsByTagName('h1')[0].textContent).toBe('LOADING')
    })

    describe('happy path', () => {
        beforeEach(() => {
            const mockJson = [
                { Description: 'METRO Blue Line', ProviderID: '8', Route: '901' },
                { Description: 'METRO Green Line', ProviderID: '8', Route: '902' },
                { Description: 'METRO Red Line', ProviderID: '9', Route: '903' }
            ]
            global.testUtil.mockFetch(true, mockJson)
            render()
        })

        it('should render the transit routes after fetching the data', done => {
            setTimeout(() => {
                const dropdown = document.getElementById('transit-route-dropdown')
                expect(dropdown.options[dropdown.selectedIndex].text).toBe(
                    'METRO Blue Line'
                )
                expect(document.body).toMatchSnapshot()
                done()
            })
        })

        it('should change to right option', done => {
            setTimeout(() => {
                const dropdown = document.getElementById('transit-route-dropdown')
                act(() => {
                    dropdown.value = '903'
                    global.testUtil.fireChange(dropdown)
                })

                expect(dropdown.options[dropdown.selectedIndex].text).toBe(
                    'METRO Red Line'
                )
                done()
            })
        })
    })
})
