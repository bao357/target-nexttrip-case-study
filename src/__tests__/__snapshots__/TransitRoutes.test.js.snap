// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transit routes happy path should render the transit routes after fetching the data 1`] = `
<body>
  <div>
    <div
      class="transit-routes-route-page"
    >
      <div
        class="field"
      >
        <label
          class="label"
        >
          Select your transit route:
        </label>
        <div
          class="control"
        >
          <div
            class="select"
          >
            <select
              id="transit-route-dropdown"
            >
              <option
                value="901"
              >
                METRO Blue Line
              </option>
              <option
                value="902"
              >
                METRO Green Line
              </option>
              <option
                value="903"
              >
                METRO Red Line
              </option>
            </select>
          </div>
        </div>
      </div>
      <div
        class="navbar"
      >
        <p>
          <a
            class="button"
            href="/"
          >
            Back
          </a>
          <a
            class="button"
            href="/directions"
          >
            Next
          </a>
        </p>
      </div>
    </div>
  </div>
</body>
`;

exports[`Transit routes should render the Transit Route page with loading initially 1`] = `
<body>
  <div>
    <div
      class="transit-routes-route-page"
    >
      <div>
        <h1>
          LOADING
        </h1>
      </div>
      <div
        class="field"
      >
        <label
          class="label"
        >
          Select your transit route:
        </label>
        <div
          class="control"
        >
          <div
            class="select"
          >
            <select
              id="transit-route-dropdown"
            />
          </div>
        </div>
      </div>
      <div
        class="navbar"
      >
        <p>
          <a
            class="button"
            href="/transitRoutes"
          >
            Next
          </a>
        </p>
      </div>
    </div>
  </div>
</body>
`;
